// realm-graphql-client/dist/Credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Credentials = (function () {
    function Credentials() {
    }
    Credentials.facebook = function (facebookToken) {
        return {
            data: facebookToken,
            provider: "facebook",
        };
    };
    Credentials.google = function (googleToken) {
        return {
            data: googleToken,
            provider: "google",
        };
    };
    Credentials.usernamePassword = function (username, password, createUser) {
        return {
            data: username,
            provider: "password",
            user_info: {
                register: createUser || false,
                password: password,
            },
        };
    };
    Credentials.azureAD = function (adToken) {
        return {
            data: adToken,
            provider: "azuread",
        };
    };
    Credentials.jwt = function (jwtToken) {
        return {
            data: jwtToken,
            provider: "jwt",
        };
    };
    Credentials.admin = function (adminToken) {
        return {
            data: adminToken,
            provider: "__admin",
        };
    };
    Credentials.anonymous = function () {
        return {
            data: null,
            provider: "__anonymous",
        };
    };
    Credentials.nickname = function (nickname, isAdmin) {
        if (isAdmin === void 0) { isAdmin = false; }
        return {
            data: nickname,
            provider: "nickname",
            user_info: {
                is_admin: isAdmin,
            },
        };
    };
    Credentials.Custom = function (provider, token, userInfo) {
        return {
            provider: provider,
            data: token,
            user_info: userInfo,
        };
    };
    return Credentials;
}());
exports.Credentials = Credentials;
